### PR TITLE
Composed image format: C translation

### PR DESCRIPTION
**Pull request by PalumboN**
Making C translation available for new Composed image Reader/Writer.

- Refactoring code to be "Slang friendly"
- Adding special cases for STON files manipulation in Slang
- Change the "main" load image function wrote in C

The translation ~works~ finishes and generate C code, but there is still some issues:
- [ ] ~Use `sscanf_s ` instead of `snprintf` to avoid overflows~ (not working on Mac 🍎 💀)
- [x] Solve the problem of reading the end of a line in STON file (`,\n`) 
- [ ] memory allocation is broken in C code

### CODE REVIEW

File: smalltalksrc/VMMaker/AbstractComposedImageAccess.class.st [hunk 1]
```diff
@@ -1,66 +1,172 @@
...
+
+	<doNotGenerate>
+	^ (format = fieldFormat)
+		  ifTrue: [ 
+			  | dataArray |
+			  dataArray := file nextLine substrings: '#:,'.
+			  (dataArray first = '}') ifTrue: [ ^false ].
+			  varHolder1 contents: dataArray second trimLineSpaces.
+			  varHolder2 contents: dataArray third trimLineSpaces asInteger ];
+		  yourself
```
Comment (PalumboN) [inline smalltalksrc/VMMaker/AbstractComposedImageAccess.class.st]: This is a hack 👎

File: src/memoryUnix.c [hunk 2]
```diff
@@ -123,16 +120,14 @@ void* allocateJITMemory(usqInt desiredSize, usqInt desiredPosition){
...
-	}
+    char *heap    =  0;
+    sqInt   heapSize    =  0;
+    sqInt   heapLimit    =  0;
 
 	pageSize = getpagesize();
 	pageMask = ~(pageSize - 1);
 
-	heapLimit = valign(max(desiredHeapSize, 1));
+	heapLimit = valign(max(desiredHeapSize, 1)) + pageSize; // Add 1 page more just in case (G & N)
```
Comment (PalumboN) [inline src/memoryUnix.c:130]: We started to fix this with Guille, but is not working yet.

File: smalltalksrc/VMMaker/AbstractComposedImageAccess.class.st (hunk 1 above)
Comment (PalumboN) [reply smalltalksrc/VMMaker/AbstractComposedImageAccess.class.st]: Should we implement it for real? (Read based to a format)

**Comment by PalumboN** [comment]
Now in CI I have this error (about trampolines (?) ) compiling C code
```
[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/4/pipeline/210#step-223-log-1271)/builds/workspace/pharo-vm_PR-388/build-StackVM/generated/64/vm/src/gcc3x-interp.c:68322:12: error: assignment to expression with array type

   aJumpBuf = (vmCallbackContext->trampoline);
```

But I can't reproduce it locally :(

**Comment by guillep** [comment]
> Now in CI I have this error (about trampolines (?) ) compiling C code
> 
> ```
> [](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/4/pipeline/210#step-223-log-1271)/builds/workspace/pharo-vm_PR-388/build-StackVM/generated/64/vm/src/gcc3x-interp.c:68322:12: error: assignment to expression with array type
> 
>    aJumpBuf = (vmCallbackContext->trampoline);
> ```
> 
> But I can't reproduce it locally :(

Just for the record, we did. It was the non JIT VM that had a different definition of longjmp, together with the C11 upgrade.

@PalumboN 's https://github.com/pharo-project/opensmalltalk-vm/pull/388/commits/f58e728c2a633ec999b9c7937ab28702527e0fe2 should fix it

**Comment by guillep** [comment]
Now it compiles in OSX and linuex but not windows:

```
error: too many arguments to function call, expected 1, have 2
                        mkdir(imageName, 777);
```

**Comment by PalumboN** [comment]
Yes, it seems that `mkdir` function is different in Windows 😕

**Comment by PalumboN** [comment]
I think that I could solve the problems, but the CI is not building my last commit:
```
[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1305)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1306)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1307)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1308)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1309)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1310)[](https://ci.inria.fr/pharo-ci-jenkins2/blue/organizations/jenkins/pharo-vm/detail/PR-388/8/pipeline#step-31-log-1311)**IncrementalGC**

**FullGC**



stack page bytes 8192 available headroom 5576 minimum unused headroom 6000



	(Terminated: 15)

[WARN ] 2022-02-25 16:41:40.000 terminateHandler (src/debugUnix.c:98):Exiting with error code 1

script returned exit code 143
```

It filled the disk again? 😞

**Comment by guillep** [comment]
Massive !

**Pull request merged**